src/System.Text.Json/tests/Serialization/Value.ReadTests.cs

```diff
@@ -380,9 +380,36 @@ private static void AssertFloatingPointBehavior<T>(T netfxExpectedValue, T netco
         public static void LongInputString(int length)
         {
             // Verify boundary conditions in Deserialize() that inspect the size to determine allocation strategy.
-            string repeated = new string('x', length - 2);
+            DeserializeLongJsonString(length);
+        }
+
+        private const int MaxInt = int.MaxValue / MaxExpansionFactorWhileTranscoding;
+
+        // NOTE: VeryLongInputString test is constrained to run on Windows and MacOSX because it causes
+        //       problems on Linux due to the way deferred memory allocation works. On Linux, the allocation can
+        //       succeed even if there is not enough memory but then the test may get killed by the OOM killer at the
+        //       time the memory is accessed which triggers the full memory allocation.
+        [ConditionalTheory(nameof(IsX64))]
+        [PlatformSpecific(TestPlatforms.Windows | TestPlatforms.OSX)]
+        [InlineData(MaxInt - 3)]
+        [InlineData(MaxInt - 2)]
+        [InlineData(MaxInt - 1)]
+        [InlineData(MaxInt)]
+        [InlineData(MaxInt + 1)]
+        [InlineData(MaxInt + 2)]
+        [InlineData(MaxInt + 3)]
+        [OuterLoop]
+        public static void VeryLongInputString(int length)
+        {
+            // Verify that deserializer does not do any multiplication on the length
+            DeserializeLongJsonString(length);
+        }
+
+        private static void DeserializeLongJsonString(int stringLength)
+        {
+            string repeated = new string('x', stringLength - 2);
```

Comment:
nit: Given we are now new-ing up quite large strings, maybe we should avoid creating too many strings that are effectively duplicates.

Change this to something like the following (though that only works on .NET Core, so we'd have to ifdef this similar to other places and keep the original impl otherwise - see example below):
```C#
char fillChar = 'x';
string json = string.Create(stringLength, fillChar, (chars, fillChar) =>
{
    chars.Fill(fillChar);
    chars[0] = '"';
    chars[chars.Length - 1] = '"';
});
Assert.Equal(stringLength, json.Length);

string str = JsonSerializer.Deserialize<string>(json);
Assert.True(json.AsSpan(1, json.Length - 2).SequenceEqual(str));
```

Test execution time would drop significantly with this change too.

https://github.com/dotnet/corefx/blob/81c91fab23b26592334ac6ff4bef34f102e3c3ce/src/System.Text.Json/tests/Serialization/Value.ReadTests.cs#L335-L342